### PR TITLE
expression: correct overflow check on function MINUS

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -283,22 +282,22 @@ func (c *arithmeticMinusFunctionClass) getFunction(args []Expression, ctx contex
...
-		}
+
 		setFlenDecimal4Int(bf.tp, args[0].GetType(), args[1].GetType())
-		sig := &builtinArithmeticMinusIntSig{baseIntBuiltinFunc{bf}}
+		sig := &builtinArithmeticMinusIntSig{
+			baseIntBuiltinFunc: baseIntBuiltinFunc{bf},
+			isLHSUnsigned:      mysql.HasUnsignedFlag(tpA.Flag),
+			isRHSUnsigned:      mysql.HasUnsignedFlag(tpB.Flag),
+		}
+		sig.isLHSUnsigned = mysql.HasUnsignedFlag(tpA.Flag)
```
Comment (hicqu) [inline expression/builtin_arithmetic.go]: There two lines is redundant?

Comment (zz-jason) [inline expression/builtin_arithmetic.go]: removed

File: expression/integration_test.go [hunk 2]
```diff
@@ -1576,6 +1576,20 @@ func (s *testIntegrationSuite) TestArithmeticBuiltin(c *C) {
...
 	c.Assert(err, NotNil)
 	c.Assert(err.Error(), Equals, "[types:1690]BIGINT UNSIGNED value is out of range in '(test.t.a - test.t.b)'")
+	rs, err = tk.Exec("select cast(-1 as signed) - cast(-1 as unsigned);")
+	c.Assert(errors.ErrorStack(err), Equals, "")
+	c.Assert(rs, NotNil)
+	rows, err = tidb.GetRows(rs)
+	c.Assert(rows, IsNil)
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[types:1690]BIGINT UNSIGNED value is out of range in '(-1 - 18446744073709551615)'")
+	rs, err = tk.Exec("select cast(-1 as unsigned) - cast(-1 as signed);")
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: add test for 
select cast(-3 as unsigned) - cast(-1 as signed);

File: expression/builtin_arithmetic.go [hunk 3]
```diff
@@ -334,37 +335,47 @@ func (s *builtinArithmeticMinusDecimalSig) evalDecimal(row []types.Datum) (*type
...
 	if isNull || err != nil {
 		return 0, isNull, errors.Trace(err)
 	}
-	if (a > 0 && -b > math.MaxInt64-a) || (a < 0 && -b < math.MinInt64-a) {
-		return 0, true, types.ErrOverflow.GenByArgs("BIGINT", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
+
+	switch {
+	case s.isLHSUnsigned && s.isRHSUnsigned:
+		if uint64(a) < uint64(b) {
+			return 0, true, types.ErrOverflow.GenByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s - %s)", s.args[0].String(), s.args[1].String()))
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: add a flag like isOverflow instead of return for every check.

Comment (zz-jason) [inline expression/builtin_arithmetic.go]: the returned error message is not the same for all branches, maybe it's better to `return`

**Comment by XuHuaiyu** [comment]
LGTM
PTAL @zimulala @hicqu

**Comment by zimulala** [comment]
LGTM

**Pull request merged**